File: src/forthcoming/ugo/ugos.f90
   PROGRAM GALAHAD_UGO_EXAMPLE  !  GALAHAD 2.8 - 03/06/2016 AT 08:35 GMT
   USE GALAHAD_UGO_double                       ! double precision version
   USE GALAHAD_NLPT_double, ONLY: NLPT_userdata_type
   IMPLICIT NONE
   INTEGER, PARAMETER :: wp = KIND( 1.0D+0 )    ! set precision
   TYPE ( UGO_control_type ) :: control
   TYPE ( UGO_inform_type ) :: inform
   TYPE ( UGO_data_type ) :: data
   TYPE ( NLPT_userdata_type ) :: userdata
   EXTERNAL :: FGH
   REAL ( KIND = wp ) :: x_l, x_u, x, f, g, h
!  REAL ( KIND = wp ), PARAMETER :: p = 4.0_wp
   REAL ( KIND = wp ), PARAMETER :: p = 0.0_wp
   x_l = - 1.0_wp; x_u = 2.0_wp                 ! bounds on x
   ALLOCATE( userdata%real( 1 ) )               ! Allocate space for parameter
   userdata%real( 1 ) = p                       ! Record parameter, p
   CALL UGO_initialize( data, control, inform )
   control%print_level = 1
   control%maxit = 100
   control%lipschitz_estimate_used = 3
   inform%status = 1                            ! set for initial entry
   CALL UGO_solve( x_l, x_u, x, f, g, h, control, inform, data, userdata,      &
                   eval_FGH = FGH )   ! Solve problem
   IF ( inform%status == 0 ) THEN               ! Successful return
     WRITE( 6, "( ' UGO: ', I0, ' evaluations', /,                             &
    &     ' Optimal solution =', ES14.6, /,                                    &
    &     ' Optimal objective value and gradient =', 2ES14.6 )" )              &
              inform%iter, x, f, g
   ELSE                                         ! Error returns
     WRITE( 6, "( ' UGO_solve exit status = ', I6 ) " ) inform%status
   END IF
   CALL UGO_terminate( data, control, inform )  ! delete internal workspace
   END PROGRAM GALAHAD_UGO_EXAMPLE

   SUBROUTINE FGH( status, x, userdata, f, g, h )
   USE GALAHAD_NLPT_double, ONLY: NLPT_userdata_type
   INTEGER, PARAMETER :: wp = KIND( 1.0D+0 )
   INTEGER, INTENT( OUT ) :: status
   REAL ( KIND = wp ), INTENT( IN ) :: x
   REAL ( KIND = wp ), INTENT( OUT ) :: f, g, h
   TYPE ( NLPT_userdata_type ), INTENT( INOUT ) :: userdata
   REAL ( KIND = wp ) :: a
!  f = - 0.5_wp * x ** 2 + userdata%real( 1 ) * x
!  g = - x + userdata%real( 1 )
!  h = - 1.0_wp
   a = 10.0_wp
   f = x * x * cos( a*x )
   g = - a * x * x * sin( a*x ) + 2.0_wp * x * cos( a*x )
   h = - a * a* x * x * cos( a*x ) - 4.0_wp * a * x * sin( a*x ) + 2.0_wp * cos( a*x )
   status = 0
   RETURN
   END SUBROUTINE FGH

File: src/forthcoming/ugo/C/ugos.c
/* ugos.c */
/* Spec test for the UGO C interface */

#include <stdio.h>
#include <math.h>
#include "ugo.h"

struct userdata_type {
   double a;
};

// Evaluate test problem objective, first and second derivatives 
int fgh(double x, double *f, double *g, double *h, const void *userdata){
   struct userdata_type *myuserdata = (struct userdata_type *) userdata;
   double a = myuserdata->a;

   *f = x * x * cos( a*x );
   *g = - a * x * x * sin( a*x ) + 2.0 * x * cos( a*x );
   *h = - a * a* x * x * cos( a*x ) - 4.0 * a * x * sin( a*x ) 
        + 2.0 * cos( a*x );
   return 0;
}

int main(void) {

    // Derived types
    void *data;
    struct ugo_control_type control;
    struct ugo_inform_type inform;

    // Initialize UGO

    ugo_initialize( &data, &control, &inform );

    // Set user-defined control options
    control.print_level = 1;
    control.maxit = 100;
    control.lipschitz_estimate_used = 3;

    // User data
    struct userdata_type userdata;
    userdata.a = 10.0;

    // Test problem bounds
    double x_l = -1.0; 
    double x_u = 2.0;     

    // Test problem objective, gradient, Hessian values
    double x, f, g, h;
    int status;

    // import problem data
    ugo_import( &control, &data, &status, &x_l, &x_u );

    // Set for initial entry
    status = 1; 
    
    // Call UGO_solve
    ugo_solve_direct( &data, &userdata, &status, &x, &f, &g, &h, fgh );    

    // Record solution information
    ugo_information( &data, &inform, &status );
    if(inform.status == 0){ // successful return
        printf("UGO successful solve\n");
        printf("iter: %d \n", inform.iter);
        printf("x: %f \n", x);
        printf("f: %f \n", f);
        printf("g: %f \n", g);
        printf("h: %f \n", h);
        printf("f_eval: %d \n", inform.f_eval);
        printf("time: %f \n", inform.time.clock_total);
        printf("status: %d \n", inform.status);
    }else{ // error returns
        printf("UGO error in solve\n");
        printf("status: %d \n", inform.status);
    }

    // Delete internal workspace
    ugo_terminate( &data, &control, &inform );

    return 0;
}
